### PR TITLE
feat/qb1296: rec speed slice

### PR DESCRIPTION
**Pull request by bjlqsnetwork**
- qb1296: add cost time tracking for upload/download slice

### CODE REVIEW

File: pp/event/download_slice.go [hunk 1]
```diff
@@ -67,32 +96,64 @@ func ReqDownloadSlice(ctx context.Context, conn core.WriteCloser) {
...
+		utils.DebugLogf("ReqIdMap.Store <==(%v, %v)", genReqId, tkSlice)
+		downSendCostTimeMap.mux.Lock()
+		var ctStat = CostTimeStat{}
+		if val, ok := downSendCostTimeMap.dataMap.Load(rsp.TaskId + rsp.SliceInfo.SliceHash); ok {
+			ctStat = val.(CostTimeStat)
+		}
+		ctStat.PacketCount = ctStat.PacketCount + 1
+		downSendCostTimeMap.dataMap.Store(tkSliceUID, ctStat)
+		downSendCostTimeMap.mux.Unlock()
+		utils.DebugLogf("downSendPacketWgMap.Store <== K:%v, V:%v]", tkSliceUID, ctStat)
```
Comment (alexstratos) [inline pp/event/download_slice.go:144]: put it in a function

File: pp/event/download_slice.go [hunk 2]
```diff
@@ -452,3 +532,31 @@ func setDownloadSliceFail(ctx context.Context, sliceHash, taskId string, isVideo
...
+		newCostTimeStat.PacketCount = oriCostTimeStat.PacketCount - 1
+		// not counting if CostTimeStat not found from dataMap
+		if newCostTimeStat.PacketCount >= 0 {
+			downSendCostTimeMap.dataMap.Store(tkSlice.TkSliceUID, newCostTimeStat)
+		}
+		// return isDownloadFinished
+		isDownloadFinished = newCostTimeStat.PacketCount == 0
+	}
+	downSendCostTimeMap.mux.Unlock()
+	if isDownloadFinished {
```
[the remark this answers is not in the document]
Comment (bjlqsnetwork) [reply pp/event/download_slice.go:552]: yeah gonna modify it with defer

File: pp/peers/pp.go [hunk 3]
```diff
@@ -94,3 +95,37 @@ func NewServer(ctx context.Context) *PPServer {
...
+func ListenSendPacket(handler func(core.WritePacketCostTime)) {
+	for {
+		select {
+		case entry, ok := <-core.CostTimeCh:
+			if ok && entry.CostTime > 0 && entry.ReqId > 0 {
+				utils.DebugLogf("received report from WritePacket: %v", entry)
+				handler(entry)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (bjlqsnetwork) [reply pp/peers/pp.go:130]: yeah gonna add it

File: pp/event/download_slice.go (hunk 1 above)
Comment (bjlqsnetwork) [reply pp/event/download_slice.go:144]: k gonna do it

File: utils/auto_clean_map.go [hunk 4]
```diff
@@ -79,3 +79,78 @@ func (m *AutoCleanMap) pushDelete(wg *sync.WaitGroup) {
 		}
 	}()
 }
+
+type AutoCleanUnsafeMap struct {
```
Comment (alexstratos) [inline utils/auto_clean_map.go:83]: add test for the new type

File: pp/peers/pp.go [hunk 5]
```diff
@@ -94,3 +98,28 @@ func NewServer(ctx context.Context) *PPServer {
 
 	return ppServer
 }
+
+func ListenSendPacket(handler func(core.WritePacketCostTime)) {
+	quit := make(chan os.Signal, 1)
```
Comment (alexstratos) [inline pp/peers/pp.go:103]: don't add os.exit(1) out of nowhere

File: utils/auto_clean_map.go (hunk 4 above)
Comment (bjlqsnetwork) [reply utils/auto_clean_map.go:83]: k adding it

File: pp/peers/pp.go (hunk 5 above)
Comment (bjlqsnetwork) [reply pp/peers/pp.go:103]: working in [qb-1373 add proper closer functions for the pp node] to re-struct the server close

File: utils/auto_clean_map.go (hunk 4 above)
Comment (bjlqsnetwork) [reply utils/auto_clean_map.go:83]: added in new commit

**Pull request merged**